node-rti/src/ext.rs

```diff
@@ -51,6 +52,40 @@ impl ProgramStorage for ExtProgramStorage {
     }
 }
 
+impl ExtProgramStorage {
+    pub fn iter(&self) -> ExtProgramStorageIter {
+        ExtProgramStorageIter {
+            key: Some(b"g::prog".to_vec()),
+        }
+    }
+}
+
+pub struct ExtProgramStorageIter {
+    key: Option<Vec<u8>>,
+}
+
+impl Iterator for ExtProgramStorageIter {
+    type Item = Program;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(key) = &self.key {
+            let new_key = sp_io::storage::next_key(key.as_ref());
+            self.key = new_key;
+        }
+        if let Some(key) = &self.key {
```

Comment:
Functional alternative (not the best, may be better, **we can combine it with the above**):

```rust
        self.key
            .as_ref()
            .filter(|key| key.starts_with(b"g::prog::"))
            .and_then(|key| gear_common::native::get_program(ProgramId::from_slice(&key[b"g::prog::".len()..])))
```

Reply:
Originally I was trying to prevent database interactions if calling `.next` on exhausted iterator. Moving filter to `self.key` assignment did the trick and made code much cleaner. Fixed

Reply:
I've written the final code and yours is pretty similar.
```rust
        self.key = self
            .key
            .as_ref()
            .and_then(|key| sp_io::storage::next_key(key))
            .filter(|key| key.starts_with(b"g::prog::"));

        self
            .key
            .as_ref()
            .and_then(|key| gear_common::native::get_program(ProgramId::from_slice(&key[b"g::prog::".len()..])))
```